File: src/Year_2024/Dez4/P1.java
package Year_2024.Dez4;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class P1 {

    static ArrayList<char[]> list = new ArrayList<>();

    static int[] dirX = {-1, 0, 1, -1, 1, -1, 0, 1};
    static int[] dirY = {1, 1, 1, 0, 0, -1, -1, -1};

    static char[] word = {'X', 'M', 'A', 'S'};

    public static void main(String[] args) throws IOException {
        readFile();
        System.out.println(countXMas());
    }

    static void readFile() throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader("/Users/gd860/IdeaProjects/AoC24/src/Year_2024.Dez4/P1.txt"))) {
            String line;
            while ((line = reader.readLine()) != null) {
                list.add(line.toCharArray());
            }
        }
    }

    static int countXMas() {
        int result = 0;
        for (int x = 0; x < list.size(); x++) {
            for (int y = 0; y < list.get(x).length; y++) {
                for (int dir = 0; dir < 8; dir++) {
                    if (checkDir(x, y, 0, dir)) {
                        result++;
                    }
                }

            }
        }
        return result;
    }


    static boolean checkDir(int x, int y, int index, int dir) {
        if (index >= word.length) {
            return true;
        }
        if (!isValidPos(x, y)) {
            return false;
        }

        if (list.get(x)[y] == word[index]) {
            return checkDir(x + dirX[dir], y + dirY[dir], index + 1, dir);
        } else {
            return false;
        }
    }

    static boolean isValidPos(int x, int y) {
        return x >= 0 && x < list.size() && y >= 0 && y < list.get(0).length;
    }
}

File: src/Dez6/P1.kt
package Dez6

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

private fun main() {
    readFile()
    findGuard()
}

private var list = ArrayList<CharArray>()

private var dir = 0;

@Throws(IOException::class)
private fun readFile() {
    val file = File("/Users/gd860/IdeaProjects/AoC24/src/Dez6/P1.txt")
    BufferedReader(FileReader(file)).use { reader ->
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            line?.let { list.add(it.toCharArray()) }
        }
    }
}

private fun findGuard() {
    for (x in list.indices) {
        for (y in list.indices) {
            if (list[x][y] == '^') {
                calcPath(x, y)
            }
        }
    }
}

private fun calcPath(x: Int, y: Int) {
    var posX = x;
    var posY = y;

    var checkPosX = x;
    var checkPosY = y;

    var newSteps = 0;

    checkPosX--

    while (isValidPos(checkPosX, checkPosY)) {
        if (list[checkPosX][checkPosY] == '.') {
            newSteps++
            posX = checkPosX
            posY = checkPosY
            list[posX][posY] = 'X'
        } else if (list[checkPosX][checkPosY] == 'X' || list[checkPosX][checkPosY] == '^') {
            posX = checkPosX
            posY = checkPosY
        } else if (list[checkPosX][checkPosY] == '#') {
            if (dir != 3) {
                dir++;
            } else {
                dir = 0
            }
            checkPosX = posX
            checkPosY = posY
        }

        when (dir) {
            0 -> {
                checkPosX--
            }

            1 -> {
                checkPosY++
            }

            2 -> {
                checkPosX++
            }

            3 -> {
                checkPosY--
            }
        }
    }
    for (row in list) {
        for (col in row) {
            print(col)
        }
        println()
    }

    print(newSteps + 1)
}

private fun isValidPos(x: Int, y: Int): Boolean {
    return x >= 0 && x < list.size && y >= 0 && y < list[0].size
}